### PR TITLE
Filter skill calc actions

### PR DESCRIPTION
**Pull request by williammetcalf**
Added search option into the skill calculator actions list

[https://gyazo.com/bfe07338ec3d954222a1aaf6c95eb7da](https://gyazo.com/bfe07338ec3d954222a1aaf6c95eb7da)

### CODE REVIEW

**Comment by Adam-** [comment]
This needs to be rebased again. I think the calculator is going through some work atm to optimize it for speed as there are so many components. @psikoi please review when you have time.

**Comment by williammetcalf** [comment]
Alright, I'll be moving this weekend so I won't be able to do that for a few days, hopefully the plugin will be more stabilized by then

**Comment by psikoi** [comment]
I reduced the load on the skill calcs, every other change to the client on this subject will be outside the scope of the skill calc plugin, so you can continue to develop and there shouldn't be conflicts.

So please rebase and fix conflicts, and also clean up your commits please. Try not to have commits to fix other commits. Tag me when you're all set so I can review :)

**Comment by deathbeam** [comment]
Needs rebase

**Comment by williammetcalf** [comment]
@psikoi rebased with new updates, sorry for the wait

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 1]
```diff
@@ -98,6 +100,7 @@
 		uiInput.uiFieldTargetXP.addActionListener(e -> onFieldTargetXPUpdated());
 	}
 
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Remove this empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 2]
```diff
@@ -66,7 +66,9 @@
 	private CacheSkillData cacheSkillData = new CacheSkillData();
 
 	private UICombinedActionSlot combinedActionSlot = new UICombinedActionSlot();
+	private UIActionFilterSlot filterActionsSlot;
 	private ArrayList<UIActionSlot> combinedActionSlots = new ArrayList<>();
+	private List<Component> uiActionSlotBoxes = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Make this final and move it above non-final declarations

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 3]
```diff
@@ -121,6 +124,17 @@ void openCalculator(CalculatorType calculatorType)
...
+
+		// Add the action filter slot
+		filterActionsSlot = new UIActionFilterSlot((filter) ->
+		{
+			renderActionSlots();
+			calculate();
+		});
+		add(filterActionsSlot);
+
+
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Remove this empty line

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 4]
```diff
@@ -121,6 +124,17 @@ void openCalculator(CalculatorType calculatorType)
...
 
+		add(Box.createRigidArea(new Dimension(0, 5)));
+
+		// Add the action filter slot
+		filterActionsSlot = new UIActionFilterSlot((filter) ->
+		{
+			renderActionSlots();
+			calculate();
+		});
+		add(filterActionsSlot);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Add empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 5]
```diff
@@ -199,34 +213,45 @@ private void renderBonusOptions()
 	private void renderActionSlots()
 	{
 		// Wipe the list of references to the slot components.
+		Stream.concat(uiActionSlotBoxes.stream(), uiActionSlots.stream())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: What about having parent component and calling removeAll? this seems really suboptimal

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 6]
```diff
@@ -199,34 +213,45 @@ private void renderBonusOptions()
...
+				UIActionSlot slot = new UIActionSlot(action);
+				uiActionSlots.add(slot); // Keep our own reference.
+//				Component box = Box.createRigidArea(new Dimension(0, 5));
+//				uiActionSlotBoxes.add(box); // So we can remove it on filter
 
-					if (slot.isSelected())
-						combinedActionSlots.remove(slot);
-					else
-						combinedActionSlots.add(slot);
+//				add(box);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Remove commented out code

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 7]
```diff
@@ -199,34 +213,45 @@ private void renderBonusOptions()
...
 			{
-				@Override
-				public void mousePressed(MouseEvent e)
-				{
-					if (!e.isShiftDown())
-						clearCombinedSlots();
+				UIActionSlot slot = new UIActionSlot(action);
+				uiActionSlots.add(slot); // Keep our own reference.
+//				Component box = Box.createRigidArea(new Dimension(0, 5));
+//				uiActionSlotBoxes.add(box); // So we can remove it on filter
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Remove commented out code

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 8]
```diff
@@ -199,34 +213,45 @@ private void renderBonusOptions()
...
 
-					slot.setSelected(!slot.isSelected());
-					updateCombinedAction();
-				}
+				slot.addMouseListener(new MouseAdapter()
+				{
+					@Override
+					public void mousePressed(MouseEvent e)
+					{
+						if (!e.isShiftDown())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: braces are required.
{

}

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 9]
```diff
@@ -199,34 +213,45 @@ private void renderBonusOptions()
...
+							clearCombinedSlots();
+
+						if (slot.isSelected())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Braces are required

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
+package net.runelite.client.plugins.skillcalculator;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java]: Add license header

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+import net.runelite.client.ui.ColorScheme;
+import net.runelite.client.ui.components.IconTextField;
+
+import javax.imageio.ImageIO;
+import javax.swing.*;
+import java.awt.*;
+import java.io.IOException;
+import java.util.function.Consumer;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java]: Fix imports according to codestyle

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+public class UIActionFilterSlot extends JPanel
+{
+	private IconTextField uiFilterField;
+	private static final ImageIcon SEARCH_ICON;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java]: static final before instance declaration

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+			throw new RuntimeException(e);
+		}
+	}
+
+	UIActionFilterSlot(Consumer<String> onFilter)
+	{
+		setLayout(new BorderLayout(0, 0));
+		setBackground(ColorScheme.DARKER_GRAY_COLOR);
+
+		JPanel uiInfo = new JPanel(new GridLayout(1, 1));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java]: Mark this final

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+import javax.imageio.ImageIO;
+import javax.swing.*;
+import java.awt.*;
+import java.io.IOException;
+import java.util.function.Consumer;
+
+public class UIActionFilterSlot extends JPanel
+{
+	private IconTextField uiFilterField;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java]: Mark this final

**Comment by williammetcalf** [comment]
All changes have been addressed I believe. @deathbeam

**Comment by psikoi** [comment]
Mind increasing the bar's height a bit? 5px on both sides should be enough, it looks good right now, but I get itchy when things could be easily great

**Comment by williammetcalf** [comment]
like so? [https://gyazo.com/ce37c5ce6e527e5068182f12bba6dea4](https://gyazo.com/ce37c5ce6e527e5068182f12bba6dea4)

**Comment by psikoi** [comment]
Maybe 5px was too much, try 3

**Comment by williammetcalf** [comment]
[https://gyazo.com/c4827fde4b25e24a5076954edfa7abd1](https://gyazo.com/c4827fde4b25e24a5076954edfa7abd1)

**Comment by psikoi** [comment]
Beautiful

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 15]
```diff
@@ -7,10 +7,10 @@
  * modification, are permitted provided that the following conditions are met:
  *
  * 1. Redistributions of source code must retain the above copyright notice, this
- *    list of conditions and the following disclaimer.
+ *	list of conditions and the following disclaimer.
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Remove this unecessary new spaces

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 16]
```diff
@@ -25,14 +25,13 @@
  */
 package net.runelite.client.plugins.skillcalculator;
 
-import java.awt.BorderLayout;
-import java.awt.Color;
-import java.awt.Dimension;
+import java.awt.*;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Do not use star imports

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 17]
```diff
@@ -66,7 +65,9 @@
 	private CacheSkillData cacheSkillData = new CacheSkillData();
 
 	private UICombinedActionSlot combinedActionSlot = new UICombinedActionSlot();
+	private UIActionFilterSlot filterActionsSlot;
 	private ArrayList<UIActionSlot> combinedActionSlots = new ArrayList<>();
+	private final JPanel uiActionsPanel = new JPanel();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Move final above non-final

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 18]
```diff
@@ -66,7 +65,9 @@
 	private CacheSkillData cacheSkillData = new CacheSkillData();
 
 	private UICombinedActionSlot combinedActionSlot = new UICombinedActionSlot();
+	private UIActionFilterSlot filterActionsSlot;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Why is this not final

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 19]
```diff
@@ -66,7 +65,9 @@
 	private CacheSkillData cacheSkillData = new CacheSkillData();
 
 	private UICombinedActionSlot combinedActionSlot = new UICombinedActionSlot();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Mark this final too when at it

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 20]
```diff
@@ -66,7 +65,9 @@
...
+	private UIActionFilterSlot filterActionsSlot;
 	private ArrayList<UIActionSlot> combinedActionSlots = new ArrayList<>();
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Mark this final, and change type to List

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 21]
```diff
@@ -199,34 +213,44 @@ private void renderBonusOptions()
...
 		// Create new components for the action slots.
-		for (SkillDataEntry action : skillData.getActions())
-		{
-			UIActionSlot slot = new UIActionSlot(action);
-			uiActionSlots.add(slot); // Keep our own reference.
-			add(slot); // Add component to the panel.
-
-			slot.addMouseListener(new MouseAdapter()
+		final String filter = filterActionsSlot.getText().toLowerCase();
+		Arrays.stream(skillData.getActions())
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Add empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 22]
```diff
@@ -199,34 +213,44 @@ private void renderBonusOptions()
...
-				public void mousePressed(MouseEvent e)
-				{
-					if (!e.isShiftDown())
-						clearCombinedSlots();
-
-					if (slot.isSelected())
-						combinedActionSlots.remove(slot);
-					else
-						combinedActionSlots.add(slot);
+				UIActionSlot slot = new UIActionSlot(action);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Mark this final

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.skillcalculator;
+
+import net.runelite.client.ui.ColorScheme;
+import net.runelite.client.ui.components.IconTextField;
+
+import javax.imageio.ImageIO;
+import javax.swing.*;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/UIActionFilterSlot.java]: The imports are still wrong

**Comment by psikoi** [comment]
I'll review and test today

**Comment by williammetcalf** [comment]
addressed further comments

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java [hunk 24]
```diff
@@ -58,16 +59,18 @@
 	static SpriteManager spriteManager;
 	static ItemManager itemManager;
 
+	private final List<UIActionSlot> combinedActionSlots = new ArrayList<>();
+	private final JPanel uiActionsPanel = new JPanel();
+	private final  UICombinedActionSlot combinedActionSlot = new UICombinedActionSlot();
```
Comment (psikoi) [inline runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: There an extra space between final and UICombinedActionSlot

**Comment by psikoi** [comment]
I think the list should be updated when the text field changes and not when you press enter, this would be more in line with what the config search bar does.

File: runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java (hunk 24 above)
Comment (psikoi) [reply runelite-client/src/main/java/net/runelite/client/plugins/skillcalculator/SkillCalculator.java]: Remove it, I mean.

**Comment by williammetcalf** [comment]
[https://gyazo.com/633b9cbc4d37dec0ffaa129d9b508ba4](https://gyazo.com/633b9cbc4d37dec0ffaa129d9b508ba4) on key release now

**Comment by psikoi** [comment]
Remove that space I mentioned on the review.

**Comment by psikoi** [comment]
Also, why on release?

**Comment by williammetcalf** [comment]
Hmmm, I guess I lost that space when I rebased, and I tried onKeyTyped first, but typing "stafff" was still making it it show results (really not sure why), so I just went with released.

**Comment by rbbi** [comment]
Is this PR still active? See #4342

**Comment by psikoi** [comment]
@Nightfirecat same for this PR I believe,

It appears #4342 was merged and both did the same thing, if that's the case, it should be closed.

**Pull request closed without merging**